### PR TITLE
migration: Add 2 keepalive related cases

### PR DESCRIPTION
**Pull request by Yingshun**
Automated 2 P2P migration cases in this Patch.
This Patch checks migration behavior with keepalive protocol.

Signed-off-by: Yingshun Cui <yicui@redhat.com>

### CODE REVIEW

**Comment by Yingshun** [comment]
**Test result:**
(.libvirt-ci-venv-ci-runtest-vnYa8N) [root@host148 migration]# avocado run --vt-type libvirt --vt-machine-type q35 virsh.migrate_options_shared..p2p_migration.disable_keepalive
JOB ID     : bcc7ad6fd8da633140f9d0f3510abe6aeb842a90
JOB LOG    : /root/avocado/job-results/job-2019-07-11T17.36-bcc7ad6/job.log
 (1/3) type_specific.io-github-autotest-libvirt.virsh.migrate_options_shared.positive_test.p2p_migration.disable_keepalive.on_both.without_postcopy: PASS (353.75 s)
 (2/3) type_specific.io-github-autotest-libvirt.virsh.migrate_options_shared.positive_test.p2p_migration.disable_keepalive.on_target.without_postcopy: PASS (318.32 s)
 (3/3) type_specific.io-github-autotest-libvirt.virsh.migrate_options_shared.negative_test.p2p_migration.disable_keepalive.on_target.without_postcopy: PASS (185.87 s)
RESULTS    : PASS 3 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB TIME   : 860.31 s
(.libvirt-ci-venv-ci-runtest-vnYa8N) [root@host148 migration]#

File: libvirt/tests/src/migration/migrate_options_shared.py [hunk 1]
```diff
@@ -540,6 +565,30 @@ def kill_qemu_target():
...
+        """
+        Drop network connection from target host for a while and then recover it
+
+        :param ipaddr: Ip address of source
+        :param slee_time: The duration(in seconds) of network broken period
+        :raise: test.error if direct rule is not added or removed correctly
+        """
+        logging.debug("Start to drop network")
+
+        firewall_cmd.add_direct_rule(firewall_rule)
```
[the remark this answers is not in the document]
Comment (Yingshun) [reply libvirt/tests/src/migration/migrate_options_shared.py]: @dzhengfy  I'm using add_direct_rule method in Firewall_cmd class, not clear about your request. Can you please make it clear?

File: libvirt/tests/src/migration/migrate_options_shared.py [hunk 2]
```diff
@@ -865,13 +939,25 @@ def check_migration_res(result):
             qemu_conf = update_config_file('qemu',
                                            qemu_conf_dict,
                                            params=params)
+            # Setup qemu.conf on source only
+            qemu_conf_source_dict = params.get("qemu_conf_source_dict")
+            if qemu_conf_source_dict:
+                update_config_file('qemu', qemu_conf_source_dict,
```
[the remark this answers is not in the document]
Comment (Yingshun) [reply libvirt/tests/src/migration/migrate_options_shared.py]: Actually they are different. L939 is using for update config in remote and source and return object for recover. And L945 will recover the config in source first and than update only in source.

File: libvirt/tests/src/migration/migrate_options_shared.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (dzhengfy) [reply libvirt/tests/src/migration/migrate_options_shared.py]: @Yingshun sorry, this was overlooked by me. Please ignore it.

File: libvirt/tests/src/migration/migrate_options_shared.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (dzhengfy) [reply libvirt/tests/src/migration/migrate_options_shared.py]: @Yingshun According to our discussion, please create a new function for only updating remote configuration file in avocado-vt. Thanks.

**Comment by Yingshun** [comment]
depends on: https://github.com/avocado-framework/avocado-vt/pull/2173

File: libvirt/tests/src/migration/migrate_options_shared.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Yingshun) [reply libvirt/tests/src/migration/migrate_options_shared.py]: @dzhengfy Updated. Please help review. Thanks!

File: libvirt/tests/src/migration/migrate_options_shared.py [hunk 3]
```diff
@@ -540,6 +564,30 @@ def kill_qemu_target():
                                                  "post-copy failed"), 60):
             test.fail("vm status is expected to 'paused (post-copy failed)'")
 
+    def drop_network_connetion(sleep_time):
+        """
+        Drop network connection from target host for a while and then recover it
+
+        :param ipaddr: Ip address of source
```
Comment (kylazhang) [inline libvirt/tests/src/migration/migrate_options_shared.py]: no this param

File: libvirt/tests/src/migration/migrate_options_shared.py [hunk 4]
```diff
@@ -540,6 +564,30 @@ def kill_qemu_target():
...
+        :param slee_time: The duration(in seconds) of network broken period
```
Comment (kylazhang) [inline libvirt/tests/src/migration/migrate_options_shared.py]: typo

File: libvirt/tests/src/migration/migrate_options_shared.py [hunk 5]
```diff
@@ -540,6 +564,30 @@ def kill_qemu_target():
                                                  "post-copy failed"), 60):
             test.fail("vm status is expected to 'paused (post-copy failed)'")
 
+    def drop_network_connetion(sleep_time):
```
Comment (kylazhang) [inline libvirt/tests/src/migration/migrate_options_shared.py]: I think it is better to change the sleep time to block time which is more clear for its means.

Comment (Yingshun) [reply libvirt/tests/src/migration/migrate_options_shared.py]: @kylazhang  Updated.

File: libvirt/tests/src/migration/migrate_options_shared.py (hunk 3 above)
Comment (Yingshun) [reply libvirt/tests/src/migration/migrate_options_shared.py]: @kylazhang Removed

File: libvirt/tests/src/migration/migrate_options_shared.py [hunk 6]
```diff
@@ -865,13 +919,23 @@ def check_migration_res(result):
...
+            if qemu_conf_dest_dict:
+                qemuDestConf = libvirt_remote.update_remote_file(params,
+                                                                 qemu_conf_dest_dict,
+                                                                 '/etc/libvirt/qemu.conf')
         # Setup libvirtd
         libvirtd_conf_dict = params.get("libvirtd_conf_dict")
         if libvirtd_conf_dict:
             libvirtd_conf = update_config_file('libvirtd',
                                                libvirtd_conf_dict,
                                                params=params)
```
Comment (dzhengfy) [inline libvirt/tests/src/migration/migrate_options_shared.py]: Same as above

File: libvirt/tests/src/migration/migrate_options_shared.py [hunk 7]
```diff
@@ -865,13 +919,23 @@ def check_migration_res(result):
             qemu_conf = update_config_file('qemu',
                                            qemu_conf_dict,
                                            params=params)
```
[the remark this answers is not in the document]
Comment (Yingshun) [reply libvirt/tests/src/migration/migrate_options_shared.py]: @dzhengfy  I keep it for other cases, some need to update it on both source and target.

**Comment by Yingshun** [comment]
depends on: https://github.com/avocado-framework/avocado-vt/pull/2183

File: libvirt/tests/src/migration/migrate_options_shared.py [hunk 8]
```diff
@@ -540,6 +542,29 @@ def kill_qemu_target():
...
+        :param block_time: The duration(in seconds) of network broken period
+        :raise: test.error if direct rule is not added or removed correctly
+        """
+        logging.debug("Start to drop network")
+
+        firewall_cmd.add_direct_rule(firewall_rule)
+        direct_rules = firewall_cmd.get(key="all-rules", is_direct=True)
+        cmdRes = re.findall(firewall_rule, direct_rules)
+        if len(cmdRes) == 0:
+            test.error("Rule '%s' is not added", firewall_rule)
```
Comment (dzhengfy) [inline libvirt/tests/src/migration/migrate_options_shared.py]: test.error("Rule '%s' is not added" % firewall_rule)

File: libvirt/tests/src/migration/migrate_options_shared.py [hunk 9]
```diff
@@ -540,6 +542,29 @@ def kill_qemu_target():
...
+
+        logging.debug("Sleep %d seconds" % int(block_time))
+        time.sleep(int(block_time))
+        firewall_cmd.remove_direct_rule(firewall_rule)
+        direct_rules = firewall_cmd.get(key="all-rules", is_direct=True)
+        cmdRes = re.findall(firewall_rule, direct_rules)
+        if len(cmdRes):
+            test.error("Rule '%s' is not removed correctly", firewall_rule)
```
Comment (dzhengfy) [inline libvirt/tests/src/migration/migrate_options_shared.py]: test.error("Rule '%s' is not removed correctly" % firewall_rule)

Comment (Yingshun) [reply libvirt/tests/src/migration/migrate_options_shared.py]: Updated.

File: libvirt/tests/src/migration/migrate_options_shared.py (hunk 8 above)
Comment (Yingshun) [reply libvirt/tests/src/migration/migrate_options_shared.py]: @dzhengfy Updated.

**Pull request merged**